### PR TITLE
Add llamav2 fp8 inference

### PR DESCRIPTION
**Pull request by bgoldberg-habana**
Add llamav2 fp8 inference
adding kv cache in fp8 option

### CODE REVIEW

File: examples/text-generation/run_generation.py
Comment (regisss) [inline examples/text-generation/run_generation.py:214]: Having 2 arguments seems a bit redundant, can't we keep only `--kv_cache_fp8`? Or do they have different meaning?

File: examples/text-generation/README.md
Comment (regisss) [inline examples/text-generation/README.md:220]:
```suggestion

To use Llama2-7b in FP8, use these instructions to enable FP8 in SynapseAI: https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Inference_Using_FP8.html
Add the `--kv_cache_fp8` argument to run the model with a KV cache allocated in fp8.
```

**Review by regisss — COMMENTED** [review]
I left a couple of comments.
Additional question: does this only work with the 7B-parameter checkpoint of Llama2 or are the 13B/70B checkpoints compatible too?

File: examples/text-generation/run_generation.py
Comment (bgoldberg-habana) [reply examples/text-generation/run_generation.py:214]: they have different meaning,
enable_quantization - enable syanpse quantization in the bridge, setting all gemm operations to fp8.
kv_cache_fp8 - is only to store kv cache in fp8 - should be used when enable_quantization is turn on but it's not a must. it's saves memory and has a better perf.

File: examples/text-generation/README.md
Comment (bgoldberg-habana) [reply examples/text-generation/README.md:220]: with your suggestion, it seems like they need to do the instructions to enable fp8.
but we already done that.. they only need to run with the flag..

will fix this - 
Add the `--kv_cache_fp8` argument to run the model with a KV cache allocated in fp8.

**Comment by bgoldberg-habana** [comment]
> I left a couple of comments.
> Additional question: does this only work with the 7B-parameter checkpoint of Llama2 or are the 13B/70B checkpoints compatible too?

It should work but we didn't commit to it.

File: examples/text-generation/run_generation.py
Comment (regisss) [reply examples/text-generation/run_generation.py:214]: Okay I see, in that case I would rename `enable_quantization` by `fp8`. Or will it also enable fp4 in the future?

File: examples/text-generation/README.md
Comment (regisss) [reply examples/text-generation/README.md]: I'm still not convinced by the phrasing here. It makes it sound like the user should visit the Gaudi doc to make it work, which is not the case. We should rather have something like: "More information here: [link]".
Also, I would add that `--enable quantization` enables to run matrix multiplications in fp8.

Comment (bgoldberg-habana) [reply examples/text-generation/README.md]: fixed.

File: examples/text-generation/run_generation.py
Comment (bgoldberg-habana) [reply examples/text-generation/run_generation.py:214]: let's stick with enable_quantization

Comment (regisss) [reply examples/text-generation/run_generation.py:214]: IMO `enable_quantization` is misleading as there are tons of different quantization format and strategies available in the HF ecosystem. People may think that int8 or fp4 quantization is available for instance.
We also use "fp8" in Accelerate for instance: https://huggingface.co/docs/accelerate/v0.24.0/en/package_reference/kwargs#accelerate.utils.FP8RecipeKwargs
So, unless there is a strong argument for sticking to `enable_quantization`, I would prefer to have a more explicit name.

Comment (bgoldberg-habana) [reply examples/text-generation/run_generation.py:214]: fixed

File: examples/text-generation/README.md
Comment (regisss) [inline examples/text-generation/README.md:222]:
```suggestion

Llama2-7b in FP8 is enabled. Use `--enable_synapse_quantization` to enable quantization in fp8.
Add the `--kv_cache_fp8` argument to run the model with a KV cache allocated in fp8.
More information on enabling fp8 in SynapseAI is available here:
https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Inference_Using_FP8.html
```

**Review by regisss — COMMENTED** [review]
Still not convinced by this arg name but anyway we can rediscuss that later.
I left a small nit and then I'll wait for the 1.13 AMI to be released on AWS to quickly test it before merging.

File: examples/text-generation/run_generation.py
Comment (polisettyvarma) [inline examples/text-generation/run_generation.py:214]: how about enable_fp8_quantization ?

Comment (regisss) [reply examples/text-generation/run_generation.py:214]: Yes, it's more explicit. Or less verbose `--fp8_quantization`, or even `--fp8` that I suggested before (but that's okay if you want to keep the term "quantization").
Ideally, argument names should be self-explanatory and as little verbose as possible, which is why I proposed changing it.

Comment (bgoldberg-habana) [reply examples/text-generation/run_generation.py:214]: but we are going to enable different quantization approach in the next sprint so we can differ between synapse quantization and Habana quantization tool (this is the name of the new tool.. )

Comment (bgoldberg-habana) [reply examples/text-generation/run_generation.py:214]: yes and the synapse one will be deprecated when we move to 1.14

Comment (regisss) [reply examples/text-generation/run_generation.py:214]: Okay I see. Would it be acceptable to have an arg name that enables synapse fp8 quantization now with 1.13 and then the same arg that enables the other tool when we switch to 1.14?

File: examples/text-generation/README.md
Comment (MrGeva) [inline examples/text-generation/README.md:227]: let's talk about it. do not merge with this yet.

**Review by MrGeva — CHANGES_REQUESTED** [review]

File: examples/text-generation/run_generation.py
Comment (regisss) [reply examples/text-generation/run_generation.py]: @bgoldberg-habana @libinta Any idea regarding my question?

**Comment by MrGeva** [comment]
LGTM

File: examples/text-generation/run_generation.py
Comment (bgoldberg-habana) [reply examples/text-generation/run_generation.py]: we change the flag name to fp8

File: examples/text-generation/run_generation.py
Comment (ulivne) [inline examples/text-generation/run_generation.py:214]: I think the flag name is misleading,  it seems like a data type and not like different flow,
also in future we may be able to quantize to other data types

**Review by ulivne — CHANGES_REQUESTED** [review]

**Comment by bgoldberg-habana** [comment]
@regisss can we merge the PR?

**Comment by regisss** [comment]
> @regisss can we merge the PR?

I'm going to try it quickly with 1.13, it should be merged in a couple of hours

**Comment by regisss** [comment]
@bgoldberg-habana I'm getting this error when running the example:
```
Traceback (most recent call last):
  File "run_generation.py", line 490, in <module>
    main()
  File "run_generation.py", line 294, in main
    generate()
  File "run_generation.py", line 275, in generate
    outputs = model.generate(
  File "/usr/local/lib/python3.8/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/generation/utils.py", line 680, in generate
    unwrap_deepspeed_model(self).allocate_kv_cache(
  File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/models/llama/modeling_llama.py", line 493, in allocate_kv_cache
    self.model.allocate_kv_cache(batch_size, max_seq_len, inp_seq_len, kv_cache_fp8)
  File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/models/llama/modeling_llama.py", line 331, in allocate_kv_cache
    layer.allocate_kv_cache(batch_size, max_seq_len, inp_seq_len, kv_cache_fp8)
  File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/models/llama/modeling_llama.py", line 267, in allocate_kv_cache
    self.self_attn.allocate_kv_cache(batch_size, max_seq_len, inp_seq_len, kv_cache_fp8)
  File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/models/llama/modeling_llama.py", line 103, in allocate_kv_cache
    self.past_key = torch.zeros(key_shape, dtype=dtype, device=device)
RuntimeError: Expected habana_helpers::is_supported_type(type) to be true, but got false.
```
It seems to be an issue with `torch.float8_e4m3fn` taht is not recognized. Do you know what could be the reason?

**Comment by bgoldberg-habana** [comment]
> @bgoldberg-habana I'm getting this error when running the example:
> 
> ```
> Traceback (most recent call last):
>   File "run_generation.py", line 490, in <module>
>     main()
>   File "run_generation.py", line 294, in main
>     generate()
>   File "run_generation.py", line 275, in generate
>     outputs = model.generate(
>   File "/usr/local/lib/python3.8/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
>     return func(*args, **kwargs)
>   File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/generation/utils.py", line 680, in generate
>     unwrap_deepspeed_model(self).allocate_kv_cache(
>   File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/models/llama/modeling_llama.py", line 493, in allocate_kv_cache
>     self.model.allocate_kv_cache(batch_size, max_seq_len, inp_seq_len, kv_cache_fp8)
>   File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/models/llama/modeling_llama.py", line 331, in allocate_kv_cache
>     layer.allocate_kv_cache(batch_size, max_seq_len, inp_seq_len, kv_cache_fp8)
>   File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/models/llama/modeling_llama.py", line 267, in allocate_kv_cache
>     self.self_attn.allocate_kv_cache(batch_size, max_seq_len, inp_seq_len, kv_cache_fp8)
>   File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/models/llama/modeling_llama.py", line 103, in allocate_kv_cache
>     self.past_key = torch.zeros(key_shape, dtype=dtype, device=device)
> RuntimeError: Expected habana_helpers::is_supported_type(type) to be true, but got false.
> ```
> 
> It seems to be an issue with `torch.float8_e4m3fn` that is not recognized. Do you know what could be the reason?

'which version do you use? 1.12? 1.13?
because it's only part of 1.13 release

**Comment by regisss** [comment]
Yes 1.13, I'm using this Docker image: vault.habana.ai/gaudi-docker/1.13.0/ubuntu20.04/habanalabs/pytorch-installer-2.1.0:latest

**Comment by regisss** [comment]
And
```
HL-SMI Version:                           hl-1.13.0-rc-fw-47.0.0.0
Driver Version:                                     1.13.0-ee32e42
```
from `hl-smi`

**Comment by bgoldberg-habana** [comment]
> is_supported_type

running on gaudi1 or 2?

**Comment by regisss** [comment]
Ah that must be this! Gaudi1 (DL1 instance on AWS), my Gaudi2 instance won't get 1.13 before next week...
fp8 is only supported on Gaudi2 right?

**Comment by regisss** [comment]
Okay let's merge then, could you just let me know the throughput you get with the example command line so that I can compare with my runs next week please?

**Review by regisss — APPROVED** [review]

**Comment by bgoldberg-habana** [comment]
i'll send you all the numbers i  have through @libinta

**Comment by regisss** [comment]
Thanks!

**Comment by regisss** [comment]
@bgoldberg-habana When I try the command of the README with Gaudi2 on 1.13, I get the following error:
```
Traceback (most recent call last):
  File "run_generation.py", line 490, in <module>
    main()
  File "run_generation.py", line 294, in main
    generate()
  File "run_generation.py", line 275, in generate
    outputs = model.generate(
  File "/usr/local/lib/python3.8/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/generation/utils.py", line 800, in generate
    return self.greedy_search(
  File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/generation/utils.py", line 1371, in greedy_search
    model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
  File "/usr/local/lib/python3.8/dist-packages/optimum/habana/transformers/models/llama/modeling_llama.py", line 584, in prepare_inputs_for_generation
    input_ids = input_ids[:, :token_idx]
RuntimeError: Graph compile failed. synStatus=synStatus 26 [Generice failure].
```

Does it work on your side?

**Comment by regisss** [comment]
@bgoldberg-habana It's actually just a `\` missing at the end of the line where the env variables are declared. Fixed in #586.

**Pull request merged**
